File: src/matcher.py
from pathlib import Path
import re

class Matcher:
    def __init__(self, path: Path) -> None:
        self.words = path.read_text().splitlines()

    def matches(self, regex: str) -> tuple[str, ...]:
        return tuple(word for word in self.words if re.fullmatch(regex, word))

if __name__ == "__main__":
    matcher = Matcher(Path("src/valid-words.txt"))
    print(f"Loaded {len(matcher.words)} words")

File: src/main/kotlin/waffle/engine/Matcher.kt
package waffle.engine

object Matcher {
    // File is loaded lazily the first time 'wordList' is accessed
    private val wordList: List<String> by lazy {
        val inputStream = this::class.java.classLoader.getResourceAsStream("valid-words.txt")
            ?: throw IllegalStateException("Resource 'valid-words.txt' not found on the classpath")

        inputStream.bufferedReader().use { reader ->
            reader.readLines()
        }
    }

    /**
     * Searches the loaded word list using the supplied regular expression.
     */
    fun search(regex: Regex): List<String> {
        return wordList.filter { regex.containsMatchIn(it) }
    }

    fun search(regex: String): List<String> = search(regex.toRegex())
}
